homeassistant/components/elmax/config_flow.py

```diff
@@ -0,0 +1,251 @@
+"""Config flow for elmax-cloud integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from elmax_api.exceptions import ElmaxBadLoginError, ElmaxBadPinError, ElmaxNetworkError
+from elmax_api.http import Elmax
+from elmax_api.model.panel import PanelEntry
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.elmax.const import (
+    CONF_ELMAX_PANEL_ID,
+    CONF_ELMAX_PANEL_NAME,
+    CONF_ELMAX_PANEL_PIN,
+    CONF_ELMAX_PASSWORD,
+    CONF_ELMAX_USERNAME,
+    DOMAIN,
+)
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+
+_LOGGER = logging.getLogger(__name__)
+
+LOGIN_FORM_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ELMAX_USERNAME): str,
+        vol.Required(CONF_ELMAX_PASSWORD): str,
+    }
+)
+
+
+def _store_panel_by_name(
+    panel: PanelEntry, username: str, panel_names: dict[str, str]
+) -> None:
+    original_panel_name = panel.get_name_by_user(username=username)
+    panel_id = panel.hash
+    collisions_count = 0
+    panel_name = original_panel_name
+    while panel_name in panel_names:
+        # Handle same-name collision.
+        collisions_count += 1
+        panel_name = f"{original_panel_name} ({collisions_count})"
+    panel_names[panel_name] = panel_id
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for elmax-cloud."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize."""
+        self._client: Elmax = None
+        self._username: str = None
+        self._password: str = None
+        self._panels_schema = None
+        self._panel_names = None
+        self._reauth_username = None
+        self._reauth_panelid = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        # When invokes without parameters, show the login form.
+        if user_input is None:
+            return self.async_show_form(step_id="user", data_schema=LOGIN_FORM_SCHEMA)
+
+        errors: dict[str, str] = {}
+        username = user_input[CONF_ELMAX_USERNAME]
+        password = user_input[CONF_ELMAX_PASSWORD]
+
+        # Otherwise, it means we are handling now the "submission" of the user form.
+        # In this case, let's try to login to the Elmax cloud and retrieve the available panels.
+        try:
+            client = Elmax(username=username, password=password)
+            await client.login()
+
+            # If the login succeeded, retrieve the list of available panels.
+            panels = await client.list_control_panels()
+
+            # Filter the online panels
+            online_panels = filter(lambda x: x.online, panels)
```

Comment:
We usually prefer generator functions over filter and lambda

Reply:
Done!